### ISSUE TITLE
Per cell UI for running cells

### ISSUE THREAD

**Issue by ellisonbg** [jupyterlab/jupyterlab#2109]
Opening to track design/implementation of our plans to have per-cell UI for running cells. @cameronoelsen can you post mockups that show the "play button in the prompt"? Also, we will need to think about what this looks like for markdown cells too.

**Comment by cameronoelsen** [comment]
As I was thinking about the running cell indication, I originally started to incorporate a traditional loading wheel inside of the cell's brackets. While it was really easy to figure out what was happening, it felt really standard and I tried making a more abstract idea. Let me know what you think:

![ezgif-1-1e41106719](https://cloud.githubusercontent.com/assets/6437976/25632804/ea9bc28e-2f29-11e7-9cf7-a421fef9f9ae.gif)

**Comment by ellisonbg** [comment]
Marking this as beta - we might have time to do it and it is high value. Can be deferred if needed.

**Comment by JasonWeill** [comment]
I've been working on a visual proof of concept by modifying the `cells` package, but the functions in this package do not correspond to plugins that take a `CommandRegistry` as an argument. @fcollonval @krassowski would this be a suitable addition to the `cell-toolbar` package? The [jlab-enhanced-cell-toolbar](https://github.com/jupyterlab-contrib/jlab-enhanced-cell-toolbar/) package, which provided design inspiration for the cell toolbar, also included a per-cell "run" button near the prompt.

**Comment by krassowski** [comment]
Some discussion points from @bollwyvl @jasongrout and @williamstein on the call today:
- vertical space is more valuable than horizontal, dragging cells is an existing feature
- survey of other frontends?
- space for "kernel" actions (customization options "run cell", "run this cell and all below") vs the toolbar which is mostly about editing notebook?
- in cocalc users are used to it, having it also in lab would allow users to more easily switch frontends:
![image](https://github.com/user-attachments/assets/e3853954-ca14-4f52-86e8-96a32ceca1c9)

**Comment by ellisonbg** [comment]
I agree that vertical space is precious (compact vertical layout is important). Also, I have been concerned that the run button would reduce the drag handle size for cells. Is the thought to add a run button to the cell toolbar as shown above in the screenshot?

**Comment by danyeaw** [comment]
Other benchmarking:

Google Collab does something similar to the approach in #16602.

https://github.com/user-attachments/assets/10f96a6b-805a-4085-b5af-ae6cbb6f4b85

VSCode places the play in the same column, but above the cell number:

https://github.com/user-attachments/assets/efce2ea5-2186-436f-bd90-aa2e76964618

Kaggle Notebooks is pretty similar to Collab:

https://github.com/user-attachments/assets/a066178d-d080-4417-925e-c6b315130ceb

Marimo puts the button in the upper right of the cell:

https://github.com/user-attachments/assets/0d106bf2-f77d-433c-abd0-2470a1bce0ba

**Comment by krassowski** [comment]
> Is the thought to add a run button to the cell toolbar as shown above in the screenshot?

Yes, that's one of the options but I also suggested that we could just increase the white space to the left of the execution counter if we want to keep run button over it.

**Comment by krassowski** [comment]
On tangent, all other fronteds seem to place the cell toolbar on the border, not fully inside the cell. I bet this helps to minimise overlap with cell code (which I know folks from @ellisonbg team put a lot of effort into).
